### ISSUE TITLE
Collection equality regression with custom comparer on sequence type (2.4.2 -> 2.5.0-2.5.2)

### ISSUE THREAD

**Issue by jskeet** [xunit/xunit#2795]
I don't *believe* this is a duplicate of #2755, #2757 or #2743 (all of which are closed). Apologies if it *does* turn out to be a duplicate.

Our real code is comparing lists of a fairly complex type - and that complex type itself implements `IEnumerable<T>`. We pass the lists to compare, and an equality comparer, into `Assert.Equals`. In 2.4.2, I believe the only comparison was the one using the equality comparer... the per-item sequence is not iterated as far as I've seen. In 2.5.0+ (including 2.5.2) it looks like the per-item sequence *is* being iterated over and the results compared - before the custom comparer is used at all.

If the element type does *not* implement `IEnumerable<>`, the test passes - the custom comparer is used, and all is well.

Repro, using .NET 6.0 (without any implicit using directives, to avoid any ambiguity):

```csharp
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace XunitIssue;

public class ComparisonTest
{
    [Fact]
    public void Test()
    {
        List<Foo> actual = new List<Foo> { new(0), new(2) };
        List<Foo> expected = new List<Foo> { new(1), new(3) };
        Assert.Equal(expected, actual, new FooDivideByTwoEqualityComparer());
    }

    public class FooDivideByTwoEqualityComparer : IEqualityComparer<Foo>
    {
        // In xUnit 2.5.2, a breakpoint here is never hit.
        public bool Equals(Foo x, Foo y) => x.Value / 2 == y.Value / 2;

        public int GetHashCode(Foo obj) =>
            throw new NotImplementedException();
    }

    public sealed class Foo : IEnumerable<string>
    {
        public int Value { get; }
        public Foo(int value) => Value = value;

        // With this implementation, in xUnit 2.5.2, the test fails with a NotImplementedException.
        // public IEnumerator<string> GetEnumerator() => throw new NotImplementedException();

        // With this implementation, in xUnit 2.5.2, the test fails with "Assert.Equal() Failure: Collections differ"
        public IEnumerator<string> GetEnumerator() => Enumerable.Repeat("", Value).GetEnumerator();

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}
```

**Comment by bradwilson** [comment]
Definitely not a duplicate, and it's been there since I shipped 2.5.0. The logic for diving into enumerables-inside-enumerables should have been conditioned on whether or not you had a custom comparer, and I just never had a test to show the bug (until now).

Thanks!

**Comment by bradwilson** [comment]
Shipped in `2.5.3`.

**Comment by jskeet** [comment]
Fabulous, thanks :)
